modules/python/src2/cv2.cpp

```diff
@@ -284,6 +284,7 @@ static bool init_submodule(PyObject * root, const char * name, PyMethodDef * met
 static inline
 bool registerTypeInModuleScope(PyObject* module, const char* type_name, PyObject* type_obj)
 {
+    Py_INCREF(type_obj); /// Give PyModule_AddObject a reference to steal.
```

Comment:
Thanks for contribution, so as for reference counting scheme will be the following:

type is registered in module scope:
 - `PyType_FromSpecWithBases` [ref = 1]
 - `registerTypeInModuleScope`-> increment [ref = 2] -> `PyModule_AddObject`-> steals - [ref = 2]
 - `registerTypeInModuleScope`-> increment [ref = 3] -> `PyModule_AddObject`-> steals - [ref = 3]
 - Decrement [ref = 2]
 
 type is registered in class score:
 - `PyType_FromSpecWithBases` -> [ref = 1]
- `registerTypeInClassScope` -> `PyDict_SetItemString` no ref changes -> [ref = 1]
 - `registerTypeInModuleScope`-> increment [ref = 2] -> `PyModule_AddObject`-> steals [ref = 2]
 - Decrement [ref = 1] 

From the above scheme it looks that `registerTypeInClassScope` also requires `Py_INCREF`, doesn't it?
Is my scheme wrong?

Reply:
Not quite.

PyDict_SetItemString calls Py_INCREF internally, so the refcount increases at that step. That's why registerTypeInClassScope can safely borrow type_obj.

* PyType_FromSpecWithBases -> [ref = 1]
* registerTypeInClassScope -> PyDict_SetItemString (calls Py_INCREF internally) -> [ref = 2]
* registerTypeInModuleScope-> increment [ref = 3] -> PyModule_AddObject-> steals [ref = 3]
* Decrement [ref = 2]

Reply:
@spikethehobbitmage Oh, I see.

Call chain for reference [PyDict_SetItemString](https://github.com/python/cpython/blob/fd43fb69962361ce4ef8e1175fde0a32726cdf77/Objects/dictobject.c#L3912) -> [PyDict_SetItem](https://github.com/python/cpython/blob/fd43fb69962361ce4ef8e1175fde0a32726cdf77/Objects/dictobject.c#LL1896-L1896C15) -> `Py_INCREF(value)`